### ISSUE TITLE
BUG: Prefix undefined when using XPath

### ISSUE THREAD

**Issue by mielvds** [oeg-upm/morph-kgc#118]
####  What Happens?
When using an xpath that contains a prefix, morph-kgc throws an undefined prefix error.

```
ElementPathNameError: ':' operator at line 1, column 8: [err:XPST0081] prefix 'premis' is not declared
```

Are XML prefixes unsupported? Or is there a bug in the XPath execution?

#### To Reproduce

Code to run:
```python
config = """
            [DataSource1]
            mappings: ./mapping.rml.ttl
         """
g_rdflib = morph_kgc.materialize(config)
```

**./mapping.rml.ttl**

```
@prefix rr: <http://www.w3.org/ns/r2rml#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix fnml: <http://semweb.mmlab.be/ns/fnml#> .
@prefix fno: <https://w3id.org/function/ontology#> .
@prefix d2rq: <http://www.wiwiss.fu-berlin.de/suhl/bizer/D2RQ/0.1#> .
@prefix void: <http://rdfs.org/ns/void#> .
@prefix dc: <http://purl.org/dc/terms/> .
@prefix foaf: <http://xmlns.com/foaf/0.1/> .
@prefix rml: <http://semweb.mmlab.be/ns/rml#> .
@prefix ql: <http://semweb.mmlab.be/ns/ql#> .
@prefix : <http://mapping.example.com/> .
@prefix haObjId: <https://data.hetarchief.be/id/object/> .
@prefix premis: <http://www.loc.gov/premis/rdf/v3/> .

:map_premis_000 rml:logicalSource :source_000 ;
	rdf:type rr:TriplesMap ;
	rdfs:label "premis" ;
	rr:predicateObjectMap :pom_000 ;
	rr:subjectMap :s_000 .

:om_000 rdf:type rr:ObjectMap ;
	rr:constant "http://www.loc.gov/premis/rdf/v3/IntellectualEntity" ;
	rr:termType rr:IRI .

:pm_000 rdf:type rr:PredicateMap ;
	rr:constant rdf:type .

:pom_000 rdf:type rr:PredicateObjectMap ;
	rr:objectMap :om_000 ;
	rr:predicateMap :pm_000 .

:rules_000 void:exampleResource :map_premis_000 ;
	rdf:type void:Dataset .

:s_000 rdf:type rr:SubjectMap ;
	rr:template "https://data.hetarchief.be/id/object/{/premis:object/premis:objectIdentifier/premis:objectIdentifierValue/text()}" .

:source_000 rml:iterator "/premis:premis" ;
	rml:referenceFormulation ql:XPath ;
	rml:source "https://raw.githubusercontent.com/viaacode/documentation/master/assets/sip_samples/basic_deec5d89-3024-4cbd-afcd-e18af4ad33ec/data/metadata/preservation/premis.xml" ;
	rdf:type rml:LogicalSource .
```

#### Environment (please complete the following information):
 - OS: macOS 12.6  
 - Python version: 3.9
 - Morph-KGC version: 2.3.1

**Comment by ArenasGuerreroJulian** [comment]
Hi @mielvds ,

The solution to this issue has been pointed out by @brunato in https://github.com/sissaschool/elementpath/issues/53.

Probably this involves retrieving a dict with the namespaces and using [`register_namespace()`](https://docs.python.org/3/library/xml.etree.elementtree.html#xml.etree.ElementTree.register_namespace).

**Comment by mielvds** [comment]
It's strange it doesn't register the namespace declarations from the XML document. How are you parsing the XML?

Else, this would be an addition to the `.ini` config?

**Comment by ArenasGuerreroJulian** [comment]
As far as I could investigate, when parsing the xml document, it is relacing the prefix with the corresponding URI, so we cannot use the prefix in XPath. We need to explicitly maintain the manescape mapping somehow.

Another possible alternative would be to get the namespacesl and use them to expand prefixes in the XPath expression.

I dont think we need to have an addition to `.ini` since it should be possible to get a prefix dict from the xml file.

I would need to take a closer look. If you want to take a look yourself, the relevant piece of code is here https://github.com/oeg-upm/morph-kgc/blob/21de161c996ad80f0ca406aadb118522556f3d73/src/morph_kgc/data_source/data_file.py#L163

**Comment by mielvds** [comment]
Your second suggestion would be first to try, since we probably want to keep RML agnostic of this.

From what I gather, you would need to add a namespaces dict to  https://github.com/oeg-upm/morph-kgc/blob/21de161c996ad80f0ca406aadb118522556f3d73/src/morph_kgc/data_source/data_file.py#L171

which you can get from
```
import xml.etree.ElementTree as ET

my_namespaces = dict([node for _, node in ET.iterparse('file.xml',
                                                        events=['start-ns'])])

```
But not sure if that means you have to parse the file twice.
Anyway, I'll see whether I can find some time to create a PR.

**Comment by ArenasGuerreroJulian** [comment]
`iterparse` is incremental so I assume it would only read the relevant `start-ns` part of the xml file (but not sure).

Actually, from the [`elementpath`](https://elementpath.readthedocs.io/en/latest/xpath_api.html#elementpath.iter_select) docs I see that we could provide `my_namespaces` to [`iter_selecct`](https://elementpath.readthedocs.io/en/latest/xpath_api.html#elementpath.iter_select):

```
xpath_result = elementpath.iter_select(xml_root, mapping_rule['iterator'], namespaces=mynamespaces, parser=XPath3Parser) 
```

**Comment by mielvds** [comment]
Indeed! And also both the [`fromstring`](https://docs.python.org/3/library/xml.etree.elementtree.html#xml.etree.ElementTree.fromstring) and [`parse`](https://docs.python.org/3/library/xml.etree.elementtree.html#xml.etree.ElementTree.parse) methods allow passing a parser instance, so it could just be a matter of creating an [`XMLPullParser`](https://docs.python.org/3/library/xml.etree.elementtree.html#xmlpullparser-objects) instance, catching the `start-ns` and also passing it to `fromstring` and `parse`.

**Comment by mielvds** [comment]
Fixed in #121 and  #129
